src/python/pants/engine/internals/parser.py

```diff
@@ -56,38 +89,51 @@ def reset(
         defaults: BuildFileDefaultsParserState,
         dependents_rules: BuildFileDependencyRulesParserState | None,
         dependencies_rules: BuildFileDependencyRulesParserState | None,
+        env_vars: EnvironmentVars,
     ) -> None:
         self._is_bootstrap = is_bootstrap
         self._defaults = defaults
         self._dependents_rules = dependents_rules
         self._dependencies_rules = dependencies_rules
+        self._env_vars = env_vars
         self._filepath = filepath
         self._target_adaptors.clear()
 
     def add(self, target_adaptor: TargetAdaptor) -> None:
         self._target_adaptors.append(target_adaptor)
 
-    def filepath(self) -> str:
-        if self._filepath is None:
-            raise AssertionError(
-                "The BUILD file filepath was accessed before being set. This indicates a "
-                "programming error in Pants. Please file a bug report at "
-                "https://github.com/pantsbuild/pants/issues/new."
-            )
-        return self._filepath
-
     def parsed_targets(self) -> list[TargetAdaptor]:
         return list(self._target_adaptors)
 
+    def _prelude_check(self, name: str, value: T | None, closure_supported: bool = True) -> T:
+        if value is not None:
+            return value
+        note = (
+            (
+                " If used in a prelude file, it must be within a function that is called from a BUILD"
+                " file."
+            )
+            if closure_supported
+            else ""
+        )
+        raise NameError(
+            softwrap(
+                f"""
+                The BUILD file {name} may only be used in BUILD files.{note}
+                """
+            )
+        )
+
+    def filepath(self) -> str:
+        return self._prelude_check("`build_file_dir`", self._filepath)
+
     @property
     def defaults(self) -> BuildFileDefaultsParserState:
-        if self._defaults is None:
-            raise AssertionError(
-                "The BUILD file __defaults__ was accessed before being set. This indicates a "
-                "programming error in Pants. Please file a bug report at "
-                "https://github.com/pantsbuild/pants/issues/new."
-            )
-        return self._defaults
+        return self._prelude_check("`__defaults__`", self._defaults)
+
+    @property
+    def env_vars(self) -> EnvironmentVars:
+        return self._prelude_check("`env`", self._env_vars, closure_supported=False)
```

Comment:
Why isn't env supported within macros?

In the StackStorm repo, the only use I've come up with for `env` is in a macro.

Reply:
> Does support mean pants would need to parse the prelude files to extract which env vars should be globally available?

Yes, this ^^

I totally agree this is something we need to fix. What I've done here is to give proper feedback when attempted rather than just failing until that is done.

Reply:
Got it. Well, I hope we can fix that in the 2.16 series of releases.

Reply:
Fix in #18273